### PR TITLE
Add support for OTEL_BSP_EXPORT_* environmental variables

### PR DESCRIPTION
**Pull request by pellared**
Partially addresses #1453.

This feature is important so that we can reuse the SDK in https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation.

## Changes

- The `BatchExportProcessorOptions` defaults can be overridden using
  `OTEL_BSP_SCHEDULE_DELAY`, `OTEL_BSP_EXPORT_TIMEOUT`,
  `OTEL_BSP_MAX_QUEUE_SIZE`, `OTEL_BSP_MAX_EXPORT_BATCH_SIZE`
  envionmental variables as defined in the
  [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.5.0/specification/sdk-environment-variables.md#batch-span-processor).

## TODO in next PRs

- https://github.com/open-telemetry/opentelemetry-dotnet/pull/2219#discussion_r681098431
- https://github.com/open-telemetry/opentelemetry-dotnet/pull/2219#discussion_r684120391

### CODE REVIEW

File: src/OpenTelemetry/BatchExportProcessorOptions.cs
Comment (reyang) [inline src/OpenTelemetry/BatchExportProcessorOptions.cs:88]: It seems that the logic will be used in many other places. Consider extracting the common logic (e.g. BatchExportProcessor doesn't understand environment variable, it simply "retrieves" key-value pair configuration from an abstract configuration provider, and the default provider is environment variable).

File: src/OpenTelemetry/BatchExportProcessorOptions.cs
Comment (reyang) [inline src/OpenTelemetry/BatchExportProcessorOptions.cs:64]: I think this brings the question - do we expect it to be transactional? Either read all environment variables or none.

File: src/OpenTelemetry/BatchExportProcessorOptions.cs
Comment (reyang) [inline src/OpenTelemetry/BatchExportProcessorOptions.cs:26]: These are specific to batch span processors, we need some logic to make sure it is not affecting other scenario (e.g. metrics).

File: src/OpenTelemetry/BatchExportProcessorOptions.cs
Comment (pellared) [reply src/OpenTelemetry/BatchExportProcessorOptions.cs:88]: Feel a little YAGNI. I will try making a dedicated PR even though I think it will be not worth adding an additional layer of interaction.

Still, I think it may be worth to rename `TryLoadEnvVarInt` and put it into some helper class to reuse it in other places.

File: src/OpenTelemetry/BatchExportProcessorOptions.cs
Comment (pellared) [reply src/OpenTelemetry/BatchExportProcessorOptions.cs:64]: Rather not. I will change it and also fix the same issue in other places.

File: src/OpenTelemetry/BatchExportProcessorOptions.cs
Comment (pellared) [reply src/OpenTelemetry/BatchExportProcessorOptions.cs:26]: Do you have any proposal on how to approach this?

I think that the problem is that the root problem is that the .NET SDK tries to use common abstractions to implement multiple OTel specification components. E.g. we have a common exporter base class for both metrics and traces exporter. I am aware it reduces some code duplication but it tightly couples the API.

Can we use `BatchExportProcessorOptions` for tracing (to not break SemVer) and create a `BatchMetricsExportProcessorOptions` for metrics SDK?

File: src/OpenTelemetry/BatchExportProcessorOptions.cs
Comment (pellared) [reply src/OpenTelemetry/BatchExportProcessorOptions.cs]: Addressed for `BatchExportProcessorOptions` here https://github.com/open-telemetry/opentelemetry-dotnet/pull/2219/commits/f679a640dcc9e6a197fb94a7aba46575a2aa63f7

I would rather address it for `OtlpExporterOptions` and `JaegerExporterOptions` together with https://github.com/open-telemetry/opentelemetry-dotnet/pull/2219#discussion_r681164662 in a dedicated PR.

File: src/OpenTelemetry/BatchExportProcessorOptions.cs
Comment (pellared) [reply src/OpenTelemetry/BatchExportProcessorOptions.cs:79]: Would it be acceptable to use `OpenTelemetrySdkEventSource` in `OtlpExporterOptions` and `JaegerExporterOptions` instead of the currently used `OpenTelemetryProtocolExporterEventSource` and `JaegerExporterEventSource` to indicate problems when reading and parsing the env vars?

File: src/OpenTelemetry/BatchExportProcessorOptions.cs
Comment (CodeBlanch) [reply src/OpenTelemetry/BatchExportProcessorOptions.cs:26]: What if we did something like...

```csharp
        internal const string SpanMaxQueueSizeEnvVarKey = "OTEL_BSP_MAX_QUEUE_SIZE";

        internal const string SpanMaxExportBatchSizeEnvVarKey = "OTEL_BSP_MAX_EXPORT_BATCH_SIZE";

        internal const string SpanExporterTimeoutEnvVarKey = "OTEL_BSP_EXPORT_TIMEOUT";

        internal const string SpanScheduledDelayEnvVarKey = "OTEL_BSP_SCHEDULE_DELAY";

        public BatchExportProcessorOptions()
        {
            if (typeof(T) == typeof(Activity))
            {
               LoadSpanEnvVars();
            }
        }

        private void LoadSpanEnvVars()
        {
            int value;

            if (TryLoadEnvVarInt(SpanExporterTimeoutEnvVarKey, out value))
            {
                this.ExporterTimeoutMilliseconds = value;
            }

            if (TryLoadEnvVarInt(SpanMaxExportBatchSizeEnvVarKey, out value))
            {
                this.MaxExportBatchSize = value;
            }

            if (TryLoadEnvVarInt(SpanMaxQueueSizeEnvVarKey, out value))
            {
                this.MaxQueueSize = value;
            }

            if (TryLoadEnvVarInt(SpanScheduledDelayEnvVarKey, out value))
            {
                this.ScheduledDelayMilliseconds = value;
            }
        }
```

I'm not really a fan of duck typing in generics like this but trying to introduce classes into the hierarchy now will probably be a breaking change.

Comment (pellared) [reply src/OpenTelemetry/BatchExportProcessorOptions.cs:26]: @CodeBlanch thanks 👍 

> but trying to introduce classes into the hierarchy now will probably be a breaking change

That is why I suggest changing the metrics SDK as AFAIK it od not stable yet. And leave the tracing SDK as it is.

Your proposal would work but maybe it is worth to make a step back

Comment (pellared) [reply src/OpenTelemetry/BatchExportProcessorOptions.cs:26]: @CodeBlanch 
I won't be that simple 😞 
 `OtlpMetricsExporter` uses `OtlpExporterOptions` which has a `BatchExportProcessorOptions<Activity>`.

Comment (pellared) [reply src/OpenTelemetry/BatchExportProcessorOptions.cs:26]: After the SIG meeting, I have an impression that this would be the preferred way forward:
1. Create an issue that `OtlpMetricsExporter` should not reference `BatchExportProcessorOptions<Activity>`
1. Create a `BatchSpanExportProcessorOptions` as a subclass of `BatchExportProcessorOptions<Activity>`

**Comment by pellared** [comment]
@open-telemetry/dotnet-approvers PTAL

File: src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs:34]: Could you update public api as well.. (We temporarily disabled it for metrics work, so CI is not checking it.)

File: src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs
Comment (pellared) [reply src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs:24]: I think that `BatchExportActivityProcessorOptions` would be a little "more English". I have also seen `SpanProcessor` term in the [specs](https://github.com/open-telemetry/opentelemetry-specification). WDYT?

File: src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs
Comment (pellared) [reply src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs:34]: Sure. Is there any tooling to help with it? How to test it (e.g. how to run the check which is disabled)?

Comment (CodeBlanch) [reply src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs:34]: @pellared If you want to re-enable the check locally make this block...

https://github.com/open-telemetry/opentelemetry-dotnet/blob/c9c78691cec293af20972e55b661b14b126cb55a/build/Common.prod.props#L4-L15

...like this...

```xml
  <ItemGroup>
    <PackageReference Include="MinVer" Version="$(MinVerPkgVer)" Condition="'$(IntegrationBuild)' != 'true'">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.CodeAnalysis.PublicApiAnalyzers" Version="$(MicrosoftCodeAnalysisAnalyzersPkgVer)" Condition=" $(OS) == 'Windows_NT'">
      <PrivateAssets>All</PrivateAssets>
    </PackageReference>
  </ItemGroup>
```

And then you should get the warnings and code fixes to resolve them.

File: src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs
Comment (cijothomas) [reply src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs:24]: `BatchExportActivityProcessorOptions` sounds good to me.

Comment (pellared) [reply src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs]: Addressed 7dfb9354

I just noticed that we have `BatchExportActivityProcessor` which lives in the `OpenTelemetry` namespace. Does it mean that `BatchExportActivityProcessorOptions` should be placed in the same namespace (instead of `OpenTelemetry.Trace`)?

File: src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs
Comment (pellared) [reply src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs]: Thank you.

Addressed 018b6a49 

Shouldn't we remove the `.publicApi/net46` and `.publicApi/net452` directories as these .NET Framework versions are no longer supported?

**Comment by dszmigielski** [comment]
@open-telemetry/dotnet-approvers could you PTAL, so we can close this one? It's been hanging here for quite a long time now.

**Review by reyang — APPROVED** [review]
LGTM. Sorry for my delayed response, I was focusing on the metrics part recently.

File: src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs
Comment (reyang) [reply src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs]: > I just noticed that we have `BatchExportActivityProcessor` which lives in the `OpenTelemetry` namespace.

IIRC it was a mistake and when folks realized that it's already too late to correct it (unless we deprecate it or introduce some breaking changes).

> Does it mean that `BatchExportActivityProcessorOptions` should be placed in the same namespace (instead of `OpenTelemetry.Trace`)?

I vote for `OpenTelemetry.Trace` namespace, and I saw that you already did this 👍

File: src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs
Comment (CodeBlanch) [inline src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs:24]: Just throwing this out there, I don't think this _needs_ to be public. Since what is exposed in the exporter options is `BatchExportProcessorOptions<Activity>`. Reason to make it internal sealed is greater flexibility to change it in the future if needed 😄

Comment (pellared) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs:24]: I made it `public` so that it can be reused in contrib and other repositories as well.

Comment (CodeBlanch) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs:24]: That's a good point, I don't feel strongly either way. Historically we have duplicated stuff in contrib with a preference to keeping the public API minimal and the "forever-support" footprint small. Defer to @cijothomas!

If we do keep it public here, probably should add some /// comments. Maybe also mark it sealed?

File: src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs
Comment (cijothomas) [reply src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs]: > IIRC it was a mistake and when folks realized that it's already too late to correct it (unless we deprecate it or introduce some breaking changes).

Yes this is true. We can keep this in OpenTelemetry.Trace ns.

**Review by cijothomas — APPROVED** [review]

File: src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs
Comment (cijothomas) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs:24]: Okay either way - as this is doing things from spec, okay to merge as is, keeping public.

Comment (cijothomas) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs:24]: agree with making sealed.
(Will merge without it! and can do a small followup for it.. I have clicked "merge main" for this PR quite a lot but always misses the chance to merge before another PR goes in!)

File: src/OpenTelemetry/README.md
Comment (cijothomas) [inline src/OpenTelemetry/README.md:207]: might be a good idea to call out that this env variables affect *all* exporters unless individually overridden in code.

**Comment by cijothomas** [comment]
@pellared Merging this. Please take a note of couple of suggestion, and consider follow ups to address them.

**Comment by pellared** [comment]
@cijothomas @CodeBlanch I have the follow-ups very high on my TODO list. I am just overwhelmed with other stuff ATM.

File: src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs
Comment (pellared) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs:24]: @cijothomas Just to double-check. You prefer to make it sealed - right?

File: src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs
Comment (pellared) [reply src/OpenTelemetry/Trace/BatchSpanExportProcessorOptions.cs]: @cijothomas @CodeBlanch  see ⬆️

File: src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs
Comment (cijothomas) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs:24]: yes

Comment (pellared) [reply src/OpenTelemetry/Trace/BatchExportActivityProcessorOptions.cs:24]: Trying to address it here: https://github.com/open-telemetry/opentelemetry-dotnet/pull/2350

**Pull request merged**
